server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -483,44 +492,50 @@ public static String getShardStateId(IndexShard indexShard, IndexCommit snapshot
-                Map<ShardId, IndexShardSnapshotStatus> localShards = currentSnapshotShards(snapshot.snapshot());
-                if (localShards != null) {
-                    Map<ShardId, ShardSnapshotStatus> masterShards = snapshot.shards();
-                    for (Map.Entry<ShardId, IndexShardSnapshotStatus> localShard : localShards.entrySet()) {
-                        ShardId shardId = localShard.getKey();
-                        ShardSnapshotStatus masterShard = masterShards.get(shardId);
-                        if (masterShard != null && masterShard.state().completed() == false) {
-                            final IndexShardSnapshotStatus.Copy indexShardSnapshotStatus = localShard.getValue().asCopy();
-                            final Stage stage = indexShardSnapshotStatus.getStage();
-                            // Master knows about the shard and thinks it has not completed
-                            if (stage == Stage.DONE) {
-                                // but we think the shard is done - we need to make new master know that the shard is done
-                                logger.debug(
-                                    "[{}] new master thinks the shard [{}] is not completed but the shard is done locally, "
-                                        + "updating status on the master",
-                                    snapshot.snapshot(),
-                                    shardId
-                                );
-                                notifySuccessfulSnapshotShard(snapshot.snapshot(), shardId, localShard.getValue().getShardSnapshotResult());
-
-                            } else if (stage == Stage.FAILURE) {
-                                // but we think the shard failed - we need to make new master know that the shard failed
-                                logger.debug(
-                                    "[{}] new master thinks the shard [{}] is not completed but the shard failed locally, "
-                                        + "updating status on master",
-                                    snapshot.snapshot(),
-                                    shardId
-                                );
-                                notifyFailedSnapshotShard(
-                                    snapshot.snapshot(),
-                                    shardId,
-                                    indexShardSnapshotStatus.getFailure(),
-                                    localShard.getValue().generation()
-                                );
-                            }
+                final Map<ShardId, IndexShardSnapshotStatus> localShards;
+                synchronized (shardSnapshots) {
+                    final var currentLocalShards = shardSnapshots.get(snapshot.snapshot());
+                    if (currentLocalShards == null) {
+                        return;
+                    }
+                    localShards = Map.copyOf(currentLocalShards);
```

Comment:
Ok to keep hold of the mutable objects here, we're in `SnapshotShardsService` which is responsible for manipulating them.